usdot-fhwa-OPS/V2X-Hub
# 503 cda sim time sync

# PR Details
## Description


This Pull Request adds functionality to the **CDASimAdapter** to listen to time sync messages formatted as follows and forward them to other **PluginClientClockAware** (s) to update their **Clock** objects with the new time.
```
{
"seq": 1232 // sequential number of time step
"timestep": 1232123123 // timestamp in ms
}
```
Also added new Status fields for PluginClientClockAware to indicate when plugin is in **SIMULATION MODE** and what the latest timestep value it has received.
## Related Issue
#503





## Motivation and Context


By adding this functionality to the **CDASimAdapter** and updating the **PluginClientClockAware** to listen for TimeSync tmx messages we are able to develop plugins that can function both in simulation and in real world deployment by allowing them to listen to simulation time sync messages in simulation and use system time calls in real world deployment.
## How Has This Been Tested?
Due to heavy reliance on TMX Core functionality I was not able to unit test my changes. I was however able to locally integration test my changes. To test these locally do the following.
1) Clear all existing containers and volumes associated with dev container deployment
2) Deploy V2X-Hub with dev container deployment (Checkout this branch of course before doing so)
3) Enable MapPlugin which extends PluginClientClockAware.
4) Check MapPlugin Status tab to see SIMULATION MODE ON status.
5) Run script under src/v2i-hub/CDASimAdapter/scripts outside dev container to publish time sync messages
6) Check MapPlugin Status tab for Simulation Time (ms) (Should update every 5 seconds)
7) Check Kowl (Web UI at localhost:8083) topic time_sync for incoming messages on kafka (Should update every 5 s)
8) Confirm no plugins are crashing 




## Types of changes




## Checklist:




[V2XHUB Contributing Guide](https://github.com/usdot-fhwa-OPS/V2X-Hub/blob/develop/Contributing.md)
